engine/lite/embeddedDeploymentManager/src/main/scala/pl/touk/nussknacker/engine/embedded/RequestResponseDeploymentStrategy.scala

```diff
@@ -0,0 +1,106 @@
+import pl.touk.nussknacker.engine.util.config.ConfigEnrichments._
+
+import scala.collection.concurrent.TrieMap
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.util.{Failure, Success, Try}
+
+object RequestResponseDeploymentStrategy {
+
+  case class RequestResponseConfig(port: Int, interface: String = "0.0.0.0")
+
+  def apply(config: Config)(implicit as: ActorSystem): RequestResponseDeploymentStrategy = {
+    new RequestResponseDeploymentStrategy(config.rootAs[RequestResponseConfig])
+  }
+
+}
+
+class RequestResponseDeploymentStrategy(config: RequestResponseConfig)(implicit as: ActorSystem)
+  extends DeploymentStrategy with LazyLogging {
+
+  private val akkaHttpSetupTimeout = 10 seconds
+
+  private val pathToInterpreter = TrieMap[String, ScenarioInterpreter]()
+
+  private var server: ServerBinding = _
+
+  override type ScenarioInterpreter = FutureBasedRequestResponseScenarioInterpreter.InterpreterType
+
+  override def open(): Unit = {
+    logger.info(s"Serving request-response on ${config.port}")
+
+    val route = new ProcessRoute(pathToInterpreter)
+
+    import as.dispatcher
+    implicit val materializer: Materializer = Materializer(as)
+
+    server = Await.result(
+      Http().newServerAt(
+      interface = config.interface,
+      port = config.port
+    ).bind(route.route), akkaHttpSetupTimeout)
+  }
+
+  override def close(): Unit = {
+    Await.result(server.terminate(akkaHttpSetupTimeout), akkaHttpSetupTimeout)
+  }
+
+
+  override def onScenarioAdded(jobData: JobData,
+                               modelData: ModelData,
+                               parsedResolvedScenario: EspProcess,
+                               contextPreparer: LiteEngineRuntimeContextPreparer)(implicit ec: ExecutionContext): Try[InterpreterType] = {
+    import pl.touk.nussknacker.engine.requestresponse.FutureBasedRequestResponseScenarioInterpreter._
+
+    val engineWithPath = pathForScenario(jobData.metaData).product(RequestResponseEngine[Future](parsedResolvedScenario, jobData.processVersion, contextPreparer, modelData, Nil,
+      ProductionServiceInvocationCollector, ComponentUseCase.EngineRuntime))
+    engineWithPath.foreach { case (path, interpreter) =>
+      pathToInterpreter += (path -> interpreter)
+      interpreter.open()
+    }
+    engineWithPath.map(_._2).fold(errors => Failure(new IllegalArgumentException(errors.toString())), Success(_))
+  }
+
+  private def pathForScenario(metaData: MetaData): Validated[NonEmptyList[FatalUnknownError], String] = metaData.typeSpecificData match {
+    case RequestResponseMetaData(path) => Valid(path.getOrElse(metaData.id))
+    case _ => Invalid(NonEmptyList.of(FatalUnknownError(s"Wrong scenario metadata: ${metaData.typeSpecificData}")))
+  }
+
+  override def onScenarioCancelled(data: InterpreterType): Unit = {
+    pathToInterpreter.find(_._2.id == data.id).foreach {
```

Comment:
It doesn't look like it would be thread safe. Would it be?

Reply:
I refactored a bit